contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -742,51 +768,152 @@ have a built-in `nil` value.
+
+Like admission controlled defaults, these controllers may consider external
+state when deciding default values and can be disabled.  Fields which are
+initialized this way must be strictly optional.
+
+Like all controllers, care must be taken to not clobber unrelated fields or
+values (e.g. in an array).  Using one of the patch or apply mechanisms is
+recommended to facilitate composition and concurrency of controllers.
 
-For example, the scheduler sets the `pod.spec.nodeName` field after the pod is
-created.
+### What May Be Defaulted
 
-Late-initializers should only make the following types of modifications:
+All forms of defaulting should only make the following types of modifications:
  - Setting previously unset fields
  - Adding keys to maps
  - Adding values to arrays which have mergeable semantics
-(`patchStrategy:"merge"` attribute in the type definition).
-
-These conventions:
- 1. allow a user (with sufficient privilege) to override any system-default
- behaviors by setting the fields that would otherwise have been defaulted.
- 1. enables updates from users to be merged with changes made during late
-initialization, using strategic merge patch, as opposed to clobbering the
-change.
- 1. allow the component which does the late-initialization to use strategic
-merge patch, which facilitates composition and concurrency of such components.
-
-Although the apiserver Admission Control stage acts prior to object creation,
-Admission Control plugins should follow the Late Initialization conventions
-too, to allow their implementation to be later moved to a 'controller', or to
-client libraries.
+   (`patchStrategy:"merge"` attribute in the type definition).
+
+In particular we never want to change or override a value that was provided by
+the user.  If they requested something invalid, they should get an error.
+
+This:
+ 1. allows a user (with sufficient privilege) to override any system-default
+    behaviors by explicitly setting the fields that would otherwise have been
+    defaulted.
+ 1. enables updates from users to be merged with default values.
+
+### Considerations For PUT Operations
+
+Once an object has been created and defaults have been applied, it's very
+common for updates to happen over time.  Kubernetes offers several ways of
+updating an object which preserve existing values in fields other than those
+being updated (e.g. strategic merge patch and server-side apply).  There is,
+however, a less obvious way of updating objects which can have bad interactions
+with default values - PUT (aka `kubectl replace`).  While we recommend a
+read-modify-write cycle for PUTs, we know that not everyone does this and that
+even that is not a perfect solution in the face of version skew.
+
+When an object is updated with a PUT, the API server will see the "old"
+object with defaulted values and the "new" object without.  For static defaults
+this is only a problem if the CREATE and the PUT used different API versions,
```

Comment:
I think that is not only for different API versions, the ClusterIP and NodePort problem https://github.com/kubernetes/kubernetes/issues/91459 was happening in the same version,

Reply:
ACK